packages/nimble-components/src/table/specs/table-data-api.md

```diff
@@ -0,0 +1,142 @@
+# Table Data API
+
+## Problem Statement
+
+A table/data-grid component can have a variety of ways to introduce data into it. We should make it clear how we expect users to introduce data into the `nimble-table` and address the implications that has on the API.
+
+## Links To Relevant Work Items and Reference Material
+
+[FAST data-grid spec](https://github.com/microsoft/fast/blob/master/packages/web-components/fast-foundation/src/data-grid/data-grid.spec.md) (see `rowsData` in _properties:_ section)
+
+[TanStack Table API](https://tanstack.com/table/v8/docs/api/core/table)
+
+## Implementation / Design
+
+In the `nimble-table`, the data to associated with rows will be exposed through a `data` property that is an array of key/value pairs. The key/value pairs within the data must have a key of type string and a value that is of a supported type. The supported value types are: `string`, `number`, `boolean`, `Date`, `null`, and `undefined`. As implied by the set of supported value types, complex value types such as arrays or objects are not supported. The supported value types are limited to ensure data operations are fast and to eliminate the need to custom sort functions, which could hurt performance. While the types are limited, a rendered column in the table can use data from multiple keys in a row's data item. The details about the way column definitions use data from the `data` property are out of scope for this spec.
```

Comment:
I wonder if it would be worthwhile in this spec to discuss possible migration paths for existing table clients that previously provided complex data. I could also see us putting this in a migration guide or component docs that we produce later.

If you think it's valuable now, I'd suggest capturing options for how SLE custom properties could be handled. The options I can think of are:
1. provide the data as a string and create a column type that can parse that string at runtime to determine its type and can render multiple types in the same column (as described by @atmgrifter00 [in the table spec](https://github.com/ni/nimble/pull/828#discussion_r1039985185)). This is similar to what is done today so it would have a lower migration cost but worse performance.
2. change the underlying data storage to be strictly typed and/or include type information. Higher refactoring effort but better performance.

If there are other options or other ways to evaluate those options I'd be interested to learn about them. But again, your call about whether to include here or just keep in mind for later.

Reply:
I lean towards not putting that in this spec. I think that would be better in some other format, such as a migration guide as you suggested.